language/tools/resource-viewer/src/lib.rs

```diff
@@ -226,6 +225,41 @@ fn pretty_print_ability_modifiers(f: &mut Formatter, abilities: AbilitySet) -> s
     Ok(())
 }
 
+impl serde::Serialize for AnnotatedMoveStruct {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        // TODO: can we avoid Box::leak here?
```

Comment:
@tnowacki @modocache: The issue I'm running into here feels similar to what we hit/fixed in the compiler: serde wants a `static` string for the type and field names, but I can't produce one because I'm reading them from on-chain. My naive attempt at a fix was going to be something like

```
struct AnnotatedMoveStructWithStringCache {
  s: AnnotatedMovedStruct
  cache: Vector<Box<String>>
```

+ using `Box::from_raw` to populate `cache` instead of using `leak`. But wanted to ask for your advice

Reply:
I pulled down your code and confirmed that the new string interning package can achieve what you're asking for here, with a few caveats:

1. We discussed this in #8782, but currently `move_symbol_pool::Symbol::as_str` returns a `&str` with the same lifetime as the symbol, instead of a `&'static str`. The underlying `str` does indeed live for the entire duration of the program, though, so it could (and probably should) be marked static. Once that function signature is changed to return `&'static str`, it works for your use case here, by writing `Symbol::from(format!("{}", self.type_)).as_str()`. This would add the struct tag to the global string table and keep it around for the rest of the resource viewer's execution.
2. I don't truly understand why serde requires a static string reference in this method signature. https://www.joshmcguigan.com/blog/understanding-serde/ explains that the name of the struct being serialized is always known at compile-time. Does that mean that you're using `serialize_struct` here in a way that isn't intended?

Reply:
Thanks for the detailed explanation!

To explain a bit more, I my use-case here is slightly different than the compiler in that:
- I would like for the strings I create to be deallocated at the same time as the `AnnotatedMoveStruct` rather than living for the duration of the program. This is important because this code will be called from `json-rpc`, which is a long-running server.
- I'm ok with allocating the same string multiple times (don't expect a lot of duplication), though avoiding this would be a nice bonus.

> don't truly understand why serde requires a static string reference in this method signature.

I think that this is because serde's code generation for the `Serialize` traits relies on compile-time macros that need to look at field/type names, but I'm not sure.

Reply:
There's one way I can think of to get around this, although its going to require two different code paths based on [`is_human_readable`](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#method.is_human_readable).

When `is_human_readable == true` this is generally going to be the "json" case, and here we don't actually care about serializing a named struct but instead we can serialize a [map](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#tymethod.serialize_map) using the non-static field names as keys.

When `is_human_readable == false` this is generally going to be the "bcs" case, and here we don't care about struct or field names since they aren't serialized in bcs, instead we can just serialize using a [tuple](https://docs.rs/serde/1.0.127/serde/trait.Serializer.html#tymethod.serialize_tuple).

Reply:
Good idea, especially re: map in the JSON case! This type doesn't ever appear on-chain, so I'm not sure we need to do anything special for BCS, but I will do this anyway just in case.

Reply:
Right, the client could just request the raw resource instead of this rendered one in that case. I guess I was over-thinking it :)